core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -137,12 +141,36 @@ class Partition(val topic: String,
     }
   }
 
+  /**
+    * Create the future replica if 1) the current replica is not in the given log directory and 2) the future replica
+    * does not exist. This method assumes that the current replica has already been created.
+    *
+    * @param logDir log directory
+    * @return true iff the future replica is created
+    */
+  def maybeCreateFutureReplica(logDir: String): Boolean = {
+    // The readLock is needed to make sure that while the caller checks the log directory of the
+    // current replica and the existence of the future replica, no other thread can update the log directory of the
+    // current replica or remove the future replica.
+    inReadLock(leaderIsrUpdateLock) {
```

Comment:
Thanks for the review. I think we need readLock here is that the log directory of the current replica is not changed by ReplicaAltherLogDirsThread when the KafkaRequestHandler thread is checking whether it needs to create future replica (when it handles AlterReplicaLogDirsRequest).

If we don't use read lock here, the following scenario can happen:

1) Broker receives AlterReplicaLogDirsRequest
2) KafkaRequestHandler calls maybeCreateFutureReplica(). It finds that the log directory of the current replica is different from the user-specified destination log directory. So it creates future replica and will later try to add this partition ReplicaAlterLogDirsManager
3) ReplicaAltherLogDirsThread calls maybeReplaceCurrentWithFutureReplica(), updated the log directory of the current replica, removed the future replica from allReplicasMap, and removed this partition from ReplicaAlterLogDirsManager.
4) KafkaRequestHandler will now add this partition to ReplicaAlterLogDirsManager.
5) Now we have an inconsistent state where ReplicaAlterLogDirsManager has this partition but this partition does not have future replica. The ReplicaAlterLogDirsThread will see ReplicaNotAvailableException and fail.

Does this make sense?

Reply:
I guess the scenario that you are describing is that there is already an ongoing replica movement across disks at step 2. However, in that case, it seems that we will remove the partition from replicaAlterLogDirsManager and the future replica in line 591 and 592. After that point, ReplicaAlterLogDirsThread can't call maybeReplaceCurrentWithFutureReplica() any more because the additional partition state check that we added in AbstractFetcherThread, right?

Reply:
Actually, in the scenario that there is already an ongoing replica movement to a different destination log directory, the following may happen:

- ReplicaAltherLogDirsThread calls maybeReplaceCurrentWithFutureReplica(), updated the log directory of the current replica and removed the future replica from allReplicasMap()

- KafkaRequestHandler calls alterReplicaLogDirs(). Because future replica has been removed, it will not remove this partition from replicaAlterLogDirsManager. Then it will create the future replica for this partition and add fetcher for this partition.

- Then ReplicaAltherLogDirsThread will remove fetcher for this partition. This leads to an inconsistent state where we have future replica for this partition but this partition is not in replicaAlterLogDirsManager.

The main issue is that, we do not have a lock that make it an atomic operation to 1) add/remove partition from replicaAlterLogDirsManager and 2) add/remove future replica for the partition. Protect maybeCreateFutureReplica() with `leaderIsrUpdateLock` could reduce the chance of this race condition.

Reply:
@lindong28 : In the scenario that there is already an ongoing replica movement to a different destination log directory, I am not sure what you described can happen. maybeReplaceCurrentWithFutureReplica() is called inside processPartitionData() and is protected by the AbstractFetcherThread.partitionMapLock. So, while this is ongoing, no new partitions can be added to the fetcher.

The other scenario that there is already an ongoing replica movement for this same partition to the same destination log directory does make sense. So, we can keep the readLock there.

Reply:
Thanks much for the discussion!